README.md

```diff
@@ -304,9 +304,49 @@ If the `foo` command would normally return a JSON object like `{"hello": "world"
 }
 ```
 
-### Authentication
+### Authentication & Authorization
+
+See the `apikey` module for a simple example of a pre-shared key.
+
+If instead you use a third party auth system that vends tokens and want your users to be able to log in and use the API, here's an example using Auth0:
+
+```go
+func main() {
+	cli.Init(&cli.Config{
+		AppName:   "example",
+		EnvPrefix: "EXAMPLE",
+		Version:   "1.0.0",
+	})
+
+  // Auth0 requires a client ID, issuer base URL, and audience fields when
+  // requesting a token. We set these up here and use the Authorization Code
+  // with PKCE flow to log in, which opens a browser for the user to log into.
```

Comment:
Any reason to use PKCE vs Device Authorization flow as recommended [here](https://auth0.com/docs/integrations/using-auth0-to-secure-a-cli)? From the [pkce docs](https://auth0.com/docs/flows/concepts/auth-code-pkce) it looks like it was more tailored to mobile apps.

Reply:
As far as I understand it the Device Authorization flow is meant for asynchronous two-device flows where the main device has no easy way to enter text (e.g. a TV and a laptop). The TV would display a code, the user goes to the URL on their laptop or phone, enters the code, while the TV pings the server every few seconds asking if the user logged in yet.

Since we are on a device with a keyboard (no need for two devices) the simpler Authorization Code flow seemed more appropriate, and the recommendation is to use PKCE to prevent certain types of attacks where the code could be intercepted (see [Section 1 of RFC 7636](https://tools.ietf.org/html/rfc7636#section-1))

Also see https://www.oauth.com/oauth2-servers/mobile-and-native-apps/

> The current best practice is to use the Authorization Flow along with launching an external browser, in order to ensure the native app cannot modify the browser window or inspect the contents. If the service supports PKCE, then that adds a layer of security to the mobile and native app flow.

Reply:
As far as I understand it the Device Authorization flow is meant for asynchronous two-device flows where the main device has no easy way to enter text (e.g. a TV and a laptop). The TV would display a code, the user goes to the URL on their laptop or phone, enters the code, while the TV pings the server every few seconds asking if the user logged in yet.

Since we are on a device with a keyboard (no need for two devices) the simpler Authorization Code flow seemed more appropriate, and the recommendation is to use PKCE to prevent certain types of attacks where the code could be intercepted (see [Section 1 of RFC 7636](https://tools.ietf.org/html/rfc7636#section-1))

Also see https://www.oauth.com/oauth2-servers/mobile-and-native-apps/

> The current best practice is to use the Authorization Flow along with launching an external browser, in order to ensure the native app cannot modify the browser window or inspect the contents. If the service supports PKCE, then that adds a layer of security to the mobile and native app flow.

Reply:
I should note there is one use case where this becomes a bit more interesting: running the CLI on remote/headless servers. In that case we could make a small improvement by also allowing the user to paste the returned auth code back into the terminal instead of just having the server wait for a redirect.

I'll leave that case for another PR, because a simple workaround today is to set up the CLI locally, then copy over the cached refresh token to the server and things will Just Work:tm:.

Reply:
Thanks for the explanation! I do wonder why auth0 doesn't recommend pcke, but since they look very similar to me minus the need to manually go to the redirect, this looks good!